File: imdb.py
import multiprocessing
import re
import requests
import proxyscrape
from bs4 import BeautifulSoup
import sqlite3


base_imdb_url = "http://www.imdb.com"
url = base_imdb_url + "/user/ur26735780/ratings"

#proxy stuff
collector = proxyscrape.create_collector('default', 'http')  # Create a collector for http resources
proxies = collector.get_proxies({'country': 'united states'})  # Retrieve a united states proxy

proxies_dict = dict()
proxies_dict[proxy.type] = f"{proxy.type}://{proxy.host}:{proxy.port}"

#get imdb movie id from parsed html
def get_imdb_id(content_array):
    if len(content_array) > 0:
        for item in content_array:
            if hasattr(item, 'attrs'):
                if 'href' in item.attrs :
                    imdb_link = item.attrs['href']
                    if(imdb_link != None):
                        if(len(imdb_link) > 0):
                            imdb_id = re.search("tt.*/",imdb_link).group(0)
                            if imdb_id != None:
                                return imdb_id[0:len(imdb_id)-2]
                                
#process url, add movie id and my user rating to db. rest of the info will be updated on node js side.
def process_page(url):
    source_code = requests.get(url)
    plain_text = source_code.text
    soup = BeautifulSoup(plain_text, 'html.parser')

    for r in soup.findAll('div', {'class': 'lister-item-content'}):
        for rating in r.findAll('div', {'class': 'ipl-rating-star ipl-rating-star--other-user small'}):
            rating = str(rating.text)
            if rating == '':
                rating = 'NaN'

        for title in r.findAll('h3', {'class': 'lister-item-header'}):
            movie_id = get_imdb_id(title.contents)
            
        if movie_id != '':
            # TODO : Add to sqlite db
            # mydict[movie_id] = rating.replace('\n', '')
        #update db safely

while(url != ''):
    pages += 1
    source_code = requests.get(url, proxies=proxies_dict)
    plain_text = source_code.text
    soup = BeautifulSoup(plain_text, 'html.parser')
    for r in soup.findAll('a', {'class': 'flat-button lister-page-next next-page'}, href=True):
        if(r.text):
            process_page(url)
            print("Processed URL\n")
            url = base_imdb_url + r['href'] 
            print("new url - {url}\n")
        else:
            url = ''

File: imdb.js
var dotenv = require('dotenv');
const axios = require('axios')
const cheerio = require('cheerio')

dotenv.config({path: __dirname + '/variables.env'});
const baseImdbURL = "http://www.imdb.com"
var prevUrl = '';

let scrapeImdb = function get(userUrl, res) {
    var movies = [];
    url = baseImdbURL + userUrl;
    axios.get(url)
        .then(response => {
            const html = response.data;
            const $ = cheerio.load(html)

            $(".lister-item-content").each(function (i, element) {
                const $element = $(element)
                const $movieId = $element.find('h3.lister-item-header a')
                const $rating = $element.find('div.ipl-rating-star.ipl-rating-star--other-user.small span.ipl-rating-star__rating')
                const movie = {
                    id: getMovieID($movieId.attr('href')),
                    rating: $rating.text()
                }
                movies.push(movie);
            })

            var nextButton = $("a.flat-button.lister-page-next.next-page");
            if (nextButton.length) {
                nextButton.each(function (i, element) {
                    const $element = $(element)
                    const $url = $element.attr('href')
                    if ($url) {
                        prevUrl = url
                        url = $url
    
                        res(movies)
                        console.log('call another get at - ' + url)
                        get(url, res)
                    }
                })
            }
            else {
                res(movies);
            }
        })

        .catch(error => {
            console.log(error)
        })
}


function getMovieID(url) {
    var regex = new RegExp('/tt.*\/');
    var result = regex.exec(url)[0]
    return result.substr(1, result.length - 2)
}

function getMovieOMDB(movie) {
    var omdbUrl = process.env.OMDB_URL;
    var omdbApiKey = process.env.OMDB_API_KEY;
    imdbUrl = omdbUrl + movie.id + omdbApiKey;
    return axios.get(imdbUrl)
        .then(response => {
            return response.data;
        })

        .catch(error => {
            console.log(error)
        })
}

module.exports = { scrapeImdb, getMovieOMDB }